File: src/main/java/programmers/Lim43165.java
package programmers;

public class Lim43165 {
    private int answer = 0;

    public int solution(int[] numbers, int target) {
        solve(0, true, 0, numbers, target);
        solve(0, false, 0, numbers, target);

        return answer;
    }

    private void solve(int idx, boolean positive, int sum, int[] numbers, int target) {
        int result = 0;
        if (positive) {
            result = sum + numbers[idx];
        } else {
            result = sum - numbers[idx];
        }

        if (idx == numbers.length - 1) {
            if (result == target) {
                answer++;
            }
            return;
        }

        solve(idx + 1, true, result, numbers, target);
        solve(idx + 1, false, result, numbers, target);
    }
}

File: src/main/java/programmers/part2/Lim43165.kt
package programmers.part2

class Lim43165 {
    private var answer = 0

    fun solution(numbers: IntArray, target: Int): Int {
        solve(0, numbers, target, 0, true)
        solve(0, numbers, target, 0, false)

        return answer
    }

    private fun solve(idx: Int, numbers: IntArray, target: Int, sum: Int, add: Boolean) {
        val tempSum = if (add) {
            sum + numbers[idx]
        } else {
            sum - numbers[idx]
        }

        if (idx == numbers.size - 1) {
            if (tempSum == target) {
                answer++
            }
            return
        }

        solve(idx + 1, numbers, target, tempSum, true)
        solve(idx + 1, numbers, target, tempSum, false)
    }
}
